apps/prairielearn/src/tests/groupExamRolePermissions.test.ts

```diff
@@ -0,0 +1,636 @@
+      assert.isOk(questionOneFirstUserAccessResponse.ok);
+
+      // The second and third questions should be inaccessible
+      const questionTwoFirstUserAccessResponse = await fetch(questionTwoUrl);
+      assert.isNotOk(questionTwoFirstUserAccessResponse.ok);
+
+      const questionThreeFirstUserAccessResponse = await fetch(questionThreeUrl);
+      assert.isNotOk(questionThreeFirstUserAccessResponse.ok);
+
+      // Switch to third user and load first question
+      const { $: $questionOneThirdUserPage } = await switchUserAndLoadAssessment(
+        studentUsers[2],
+        questionOneUrl,
+        'attach-file-form',
+      );
+      $ = $questionOneThirdUserPage;
+
+      // The "next question" button skips unviewable questions
+      const nextQuestionLink = $('#question-nav-next').attr('href');
+      assert.strictEqual(siteUrl + nextQuestionLink, questionThreeUrl + '/');
+
+      const res = await fetch(questionThreeUrl);
+      assert.isOk(res.ok);
+      $ = cheerio.load(await res.text());
+
+      // The "previous question" button skips unviewable questions
+      const prevQuestionLink = $('#question-nav-prev').attr('href');
+      assert.strictEqual(siteUrl + prevQuestionLink, questionOneUrl + '/');
+
+      // Save and grade button is not disabled with correct permission
+      const { $: $questionOneSecondUserPage } = await switchUserAndLoadAssessment(
+        studentUsers[1],
+        questionOneUrl,
+        'attach-file-form',
+      );
+      $ = $questionOneSecondUserPage;
+
+      const secondUserButton = $('.question-grade');
+      assert.isFalse(secondUserButton.is(':disabled'));
+
+      // Switch to first user
+      const { $: $questionOneFirstUserPage, csrfToken: questionOneFirstUserCsrfToken } =
+        await switchUserAndLoadAssessment(studentUsers[0], questionOneUrl, 'attach-file-form');
+      $ = $questionOneFirstUserPage;
+
+      // Save and grade button should be disabled without correct permission
+      const firstUserButton = $('.question-grade');
+      assert.isTrue(firstUserButton.is(':disabled'));
+      const popover = $('.btn[aria-label="Locked"]');
+      assert.lengthOf(popover, 1);
+      const popoverContent = popover.data('content');
+      assert.strictEqual(
+        popoverContent,
+        'You are not assigned a role that can submit this question.',
+      );
+
+      // Get question variant
+      const questionForm = $('.question-form input[name="__variant_id"]');
+      assert.lengthOf(questionForm, 1);
+      assert.nestedProperty(questionForm[0], 'attribs.value');
+      const variantIdString = questionForm.first().attr('value');
+      assert.isDefined(variantIdString);
+      const variantId = Number.parseInt(variantIdString as string);
+
+      const variant = await queryValidatedOneRow(
+        sql.select_variant,
+        {
+          variant_id: variantId,
+        },
+        VariantIdSchema,
+      );
```

Comment:
If you already have the variant ID, why do this? AFAICT this only returns an `id` column, but you already have the ID.